opencog/atomutils/FindUtils.cc

```diff
@@ -273,44 +273,39 @@ bool contains_atomtype(const HandleSeq& clauses, Type atom_type)
 	return false;
 }
 
-
-HandleSeq get_free_vars_in_tree(const Handle& tree)
+OrderedHandleSet get_free_variables(const Handle& h, int quotation_level)
 {
-	OrderedHandleSet varset;
-
-	std::function<void (const Handle&)> find_rec = [&](const Handle& h)
-	{
-		Type t = h->getType();
-		if (t == VARIABLE_NODE)
-		{
-			varset.insert(h);
-			return;
-		}
-
-		// XXX FIXME Add support for UNQUOTE_LINK
-		if (t == QUOTE_LINK) return;
-		if (classserver().isA(t, SCOPE_LINK))
-		    // // A sugary version of ImplicationLink can be a ScopeLink
-		    // // only if it has 3 arguments
-		    // and (not classserver().isA(t, IMPLICATION_LINK)
-		    //      or h->getArity() == 3))
-			return;
-
-		if (h->isLink())
-		{
-			for (const Handle& oh : h->getOutgoingSet())
-				find_rec(oh);
-		}
-	};
-
-	find_rec(tree);
+	Type t = h->getType();
 
-	return HandleSeq(varset.begin(), varset.end());
+	// Base cases
+	if (t == VARIABLE_NODE and quotation_level < 1)
+		return {h};
+	if (h->isNode())
+		return {};
+
+	// Recursive cases
+	OrderedHandleSet results;
+	OC_ASSERT(h->isLink());
+	if (t == QUOTE_LINK) quotation_level++;
+	else if (t == UNQUOTE_LINK) quotation_level--;
+	for (const Handle& ch : h->getOutgoingSet()) {
+		OrderedHandleSet ch_free_var = get_free_variables(ch, quotation_level);
+		results.insert(ch_free_var.begin(), ch_free_var.end());
+	}
+	// If the link was a scope link then remove the scoped
+	// variables from the free variables found.
```

Comment:
This is probably OK for now, but is kind-of dangerous . bound variables in a scope link behave as if they are quoted.  So, here is a nasty example that could occur in real-life:
```
(SomeLink
    (VariableNode  $x)   ;;; -- this is free
    (ScopeLink
           (Variable $x)  ;; --- this is bound, and is not really the "same" variable as above. 
           ...))
```

Note that the second $x can be safely renamed (alpha-converted) into some other variable, leaving the first $x free.  Thus, I suggest that the correct long-term behavior is to treat the ScopeLink as if it was a QuoteLink -- any bound variables inside the ScopeLink behave as if they were quoted, from the point of view of anyone outside of the scope.

I'm not sure, I thought that there already was code somewhere else that already did something like this or almost like this.